### ISSUE TITLE
Factorize interface

### ISSUE THREAD

**Issue by emstoudenmire** [ITensor/ITensors.jl#90]
The factorize function doesn't always seem to generate a unitary U as the first return value. Hopefully I am not doing something wrong here, but here is some code to reproduce what I found:
```
l = Index(4,"Link")
s = Index(2,"Site")
T = randomITensor(l,s)
U,R = factorize(T,s)

@show U*prime(U,"Link")  # not equal to the identity
```

**Comment by mtfishman** [comment]
Hey Miles,

I explained the `factorize` function wrong to you before. What `factorize` does is the following:

```julia
i = Index(2,"i")
j = Index(2,"j")
T = randomITensor(i,j)

# Make A = U and B = S*V
A,B = factorize(T,i,dir="fromleft")

# Make A = U*S and B = V
A,B = factorize(T,i,dir="fromright")

# Make A = V*sqrt(S) and B = sqrt(S)*U
A,B = factorize(T,i,dir="center")
```

Right now, the default is "center", so you are actually getting the correct result.

Then, `factorize` also has two modes, `which_factorization = "svd"` which uses an SVD to get `U,S,V`, and the `which_factorization = "eigen"` which, in the case of "fromleft" or "fromright", uses an eigendecomposition of the density matrix to get the `U` and `S*V` factors (or `U*S` and `V` factors).

The idea is that all of these are different versions of a low rank factorization of a matrix.

Cheers,
Matt

**Comment by mtfishman** [comment]
Perhaps this is too much functionality packed into one function, and as we were discussing the "fromleft" and "fromright" may be superfluous because you could instead use the other set of indices to get the "fromright" factorization. I guess these two are essentially the same:
```julia
A,B = factorize(T,i,dir="fromright")
```
and
```julia
A,B = factorize(T,j,dir="fromleft")
```

**Comment by emstoudenmire** [comment]
Yes, let's think about this some more, because I think it could become an easier function to use if we simplified it and even split it into separate functions perhaps.

One thing to observe is that, in addition to having the freedom of which indices to put in the arguments, one can also switch the order of the result tensors on the left. So given all of that freedom already, I think the keyword arguments could be dropped.

Then to switch between the fromright/fromleft functionality (which could just be combined into one call, where you switch between these two by which indices you put or how you order the return variables), and between the "center" functionality, we could just make the "center" one its own function. 

So what do you think of that idea? Two functions like this (but with a better names):

* `Q,B = factor(T,i)  # does Q = U, B = S*V`
* `A,B = symfactor(T,i) #does A = U*sqrt(S), B = sqrt(S)*V`

Alternatively we just keep one function name `factorize` and just have a keyword argument `symmetric` which defaults to `false`?

**Comment by mtfishman** [comment]
I am fine with either separating into two different functions or having a keyword argument that switches between the current `dir="fromleft"` and `dir="center"` options (with "fromleft" being default). However, I think we should not call the keyword argument `symmetric`, since that could be conflated with the ITensor being treated as a symmetric matrix (part of why I created the `dir` keyword).

Unfortunately I have searched for names for these types of factorizations and they don't have standard names. The reason I chose `factorize` is because they are both examples of rank factorizations (https://en.wikipedia.org/wiki/Rank_factorization). What distinguishes them is that one version orthogonalizes and the other doesn't, so maybe we could do something like:

```
Q,B = factorize_ortho(T,i)   # Q = U, B = S*V
A,B = factorize(T,i)   # A = U*sqrt(S), B = sqrt(S)*V
```

so `factorize` is kind of a catch-all for "split up the ITensor with some sort of matrix decomposition", and `factorize_ortho` means "do a matrix decomposition where the first tensor is orthogonal". Better suggestions are welcome.

**Comment by orialb** [comment]
I just wanted to comment that I personally find the current `dir` keyword argument version more readable than the suggested new form where one needs to switch the order of the result tensors and the index given to `factorize` (maybe the actual keyword name could be improved).

The reason why I have easier time with the current version is that while an ITensor doesn't have a notion of index ordering, when I need to reason about what the code does I personally have to imagine (or draw) some kind of specific ordering of the tensor indices (think of a manipulation of some elements of a tensor network where there is a natural notion of left and right indices).  
So I find it clearer that the first tensor returned from `factorize` is the "left" one (and contains the provided left index `i`) and the second is the "right" one. Otherwise I have to do some kind of transpose in my head while writing code. 

But that might be just my own limitation and it might not be compatible with ITensors philosophy. 

If the keyword version is used, a suggestion for a keyword improvement is maybe use `ortho` instead of `dir`
```julia
A,B = factorize(T,i,ortho="left")  # A = U*S , B= V
A,B = factorize(T,i,ortho="right") # A = U , B= S*V
```

**Comment by mtfishman** [comment]
It is good to get feedback that choosing the direction can still be useful, I agree that it can be helpful to think in terms of left and right directions (especially with MPS). I like your idea of calling the keyword argument ortho (I guess `ortho="left"` reads as the left tensor is orthogonalized, which is easy to explain). I think the key issue here is really that the default option right now is the rarely used "center" argument, whereas it should be the `ortho="left"` option you show above. So I would propose:
```julia
A,B = factorize(T,i,ortho="left") # A = U , B = S*V, default
A,B = factorize(T,i,ortho="right")  # A = U*S , B = V
A,B = factorize(T,i,ortho="none")  # A = U*sqrt(S) , B = sqrt(S)*V, or some other factorization
```
Then, one can just use the default option all the time and use a different set if indices if they would like.

**Comment by mtfishman** [comment]
Closed by #246.
